### PR TITLE
POC for `odo push -o json` and icon animation in App Explorer

### PR DESCRIPTION
**Pull request by dgolovin**
This PR uses `odo push -o json` to push component and showes in terminal executed devfile commands and their output. Stdout output is colored black and stderror is red.

This PR uses pseudo terminal to print output to terminal view, that allows catch begining and end of command for specific node and update icon and text to different ones while command is in progress.

progress-animation

Signed-off-by: Denis Golovin dgolovin@redhat.com

### CODE REVIEW

**Comment by dgolovin** [comment]
Related to #1444.

File: src/odo.ts [hunk 1]
```diff
@@ -752,11 +765,41 @@ export class OdoImpl implements Odo {
     }
 
     public async executeInTerminal(command: string, cwd: string = process.cwd(), name = 'OpenShift'): Promise<void> {
-        const cmd = command.split(' ')[0];
-        const toolLocation = await ToolsConfig.detect(cmd);
-        const terminal: Terminal = WindowUtil.createTerminal(name, cwd);
-        terminal.sendText(toolLocation === cmd ? command : command.replace(cmd, `"${toolLocation}"`).replace(new RegExp(`&& ${cmd}`, 'g'), `&& "${toolLocation}"`), true);
+        const process = await this.spawn(command, cwd);
+        const terminal: Terminal = WindowUtil.createPseudoTerminal(name);
+        terminal.sendText('Pushing component to the cluster');
```
Comment (fbricon) [inline src/odo.ts:770]: is this function exclusively used for pushing things to the cluster?

Comment (mohitsuman) [reply src/odo.ts:770]: @dgolovin If we run any other command like "List catalog components" or "List catalog services", it sends the same text to the terminal.

File: src/odo.ts [hunk 2]
```diff
@@ -266,17 +275,19 @@ export class OpenShiftComponent extends OpenShiftObjectImpl {
         super(parent, name, contextValue, '', Collapsed, contextPath, compType, builderImage);
     }
     get iconPath(): Uri {
-        if (this.contextValue === ContextType.COMPONENT_PUSHED || this.contextValue === ContextType.COMPONENT || this.contextValue === ContextType.COMPONENT_NO_CONTEXT) {
-            let iconPath: Uri;
+        let iconPath: Uri;
+        if (this.locked) {
+            iconPath = Uri.file(path.join(__dirname, "../../images/component", 'progress.gif'));
```
Comment (mohitsuman) [inline src/odo.ts:280]: The gif images does not fit well with the design. We can use the label provided by vscode. `$(sync~spin)` to load the spinner.

File: src/odo.ts (hunk 1 above)
Comment (dgolovin) [reply src/odo.ts:770]: @fbricon yes, in this POC only. It just a test how to use pseudo terminal and show output in terminal view. It has pros and cons. With pseudo terminal we can now detect start and finish of executed process, but it comes at price of running process with spawn and basically using terminal as a view to show output. There is no way to type and execute commands in it.

File: src/odo.ts (hunk 2 above)
Comment (dgolovin) [reply src/odo.ts:280]: I would say it is perfect image for POC ;) created in 30 seconds.

**Comment by mohitsuman** [comment]
As we are moving to odo dev workflow and it opens a pseudo-terminal for it. @dgolovin we should close this PR.

**Pull request closed without merging**